crates/matrix-sdk-base/src/response_processors/state_events.rs

```diff
@@ -248,16 +307,762 @@ where
         .unzip()
 }
 
+/// Check if `m.room.create` isn't creating a loop of rooms.
+pub fn is_create_event_valid(
+    context: &mut Context,
+    room_id: &RoomId,
+    event: &SyncStateEvent<RoomCreateEventContent>,
+    state_store: &BaseStateStore,
+) -> bool {
+    let mut already_seen = BTreeSet::new();
+    already_seen.insert(room_id.to_owned());
+
+    let Some(mut predecessor_room_id) = event
+        .as_original()
+        .and_then(|event| Some(event.content.predecessor.as_ref()?.room_id.clone()))
+    else {
+        // `true` means no problem. No predecessor = no problem here.
+        return true;
+    };
+
+    loop {
+        // We must check immediately if the `predecessor_room_id` is in `already_seen`
+        // in case of a room is created and marks itself as its predecessor in a single
+        // sync.
+        if already_seen.contains(AsRef::<RoomId>::as_ref(&predecessor_room_id)) {
+            // Ahhh, there is a loop with `m.room.create` events!
+            return false;
+        }
+
+        already_seen.insert(predecessor_room_id.clone());
+
+        // Where is the predecessor room? Check in `room_infos` and then in
+        // `state_store`.
+        let Some(next_predecessor_room_id) = context
+            .state_changes
+            .room_infos
+            .get(&predecessor_room_id)
+            .and_then(|room_info| Some(room_info.create()?.predecessor.as_ref()?.room_id.clone()))
+            .or_else(|| {
+                state_store
+                    .room(&predecessor_room_id)
+                    .and_then(|room| Some(room.predecessor_room()?.room_id))
+            })
+        else {
+            // No more predecessor found. Everything seems alright. No loop.
+            break;
+        };
+
+        predecessor_room_id = next_predecessor_room_id;
+    }
```

Comment:
This seems like a fine check, though do we want to limit the loop iteration count somehow?